File: 6-kyu/standardDeviation/langs/kata.hs
-- Deviation standar of population (sample: n-1)

import Prelude

my :: [Double] -> Double
my n = (sum n) / fromIntegral (length n)

standar :: [Double] -> Double
standar n = desviation
  where
    desviation = sqrt ((sum dist) / fromIntegral (length n))
    dist = [(^) (x - mysum) 2 | x <- n]
    mysum = my n

minDeviation :: [Double] -> Double
minDeviation n = (my n) - 2.5 * (standar n)

maxDeviation :: [Double] -> Double
maxDeviation n = (my n) + 2.5 * (standar n)

recursiveFilter :: [Double] -> Int -> [Double]
recursiveFilter n y
  | y >= 1 = recursiveFilter list (y - 1)
  | y == 0 = list
  where
    list = [x | x <- n, x >= (minDeviation n), x <= (maxDeviation n)]

filterVal :: [Int] -> ((Int, Int), Double, Double)
filterVal n = ((lengOriginal, lengFilter), my list, standardDeviation)
  where
    list = recursiveFilter doubleList lengOriginal
    lengOriginal = length doubleList
    lengFilter = length list
    doubleList = map fromIntegral n
    standardDeviation = (standar list) / (fromIntegral lengFilter ** 0.5)

File: 6-kyu/standardDeviation/langs/kata.py
from math import sqrt

def filter_val(lst):
    original_length = len(lst)
    while True:
        mean = sum(lst) / len(lst)
        std_dev = sqrt(sum((x - mean) ** 2 for x in lst) / len(lst))
        len2 = [vote for vote in lst if mean - 2.5*std_dev <= vote <= mean + 2.5*std_dev]
        
        if len(len2) == len(lst):
            break
        else:
            lst = len2
    return [[original_length, len(lst)], mean, std_dev/sqrt(len(lst))]

print(filter_val([35,34,37,32,33,36,38,32,35,35,36,34,35,100,85,70]))